File: Leetcode Medium/695MaxAreaOfIsland.py
#https://leetcode.com/problems/max-area-of-island/description/
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        currCount = 0
        def helper(i, j):
            nonlocal currCount
            if grid[i][j] == 0: return
            if grid[i][j] == 1:
                currCount += 1
                grid[i][j] = 0
            if i != 0: helper(i-1, j)
            if i != len(grid)-1: helper(i+1, j)
            if j != 0: helper(i, j-1)
            if j != len(grid[0])-1: helper(i, j+1)
        
        
        
        maxRet = 0
        for i, row in enumerate(grid):
            for j, entry in enumerate(row):
                if grid[i][j] == 1:
                    helper(i, j)
                    maxRet = max(maxRet, currCount)
                    currCount = 0
        return maxRet

File: C++/Leetcode Medium/695MaxAreaOfIsland.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int res = 0;
        for (int i = 0; i < grid.size(); ++i){
            for (int j = 0; j < grid[0].size(); ++j){
                res = max(res, helper(i, j, grid));
            }
        }
        return res;
    }
    int helper(int i, int j, vector<vector<int>>& grid){
        if (i < 0 || i >= grid.size()) return 0;
        if (j < 0 || j >= grid[0].size()) return 0;
        if (grid[i][j] == 0) return 0;
        int res = 0;
        ++res;
        grid[i][j] = 0;
        res += helper(i-1, j, grid); 
        res += helper(i, j-1, grid); 
        res += helper(i+1, j, grid); 
        res += helper(i, j+1, grid); 
        return res;
    }
};